2.0/m_rpg.cpp

```diff
@@ -0,0 +1,374 @@
+			{
+				if (!c->HasUser(user))
+				{
+					user->WriteNumeric(ERR_NOTONCHANNEL, "%s %s :You are not on that channel!", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+				if (c->GetPrefixValue(user) < OP_VALUE)
+				{
+					user->WriteNumeric(ERR_CHANOPRIVSNEEDED, "%s %s :You're not a channel operator", user->nick.c_str(), parameters[0].c_str());
+					return CMD_FAILURE;
+				}
+			}
+			else
+			{
+				user->WriteNumeric(ERR_NOSUCHCHANNEL, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+				return CMD_FAILURE;
+			}
+		}
+		else
+			this->text = parameters[1];
+
+		/* Source is in the form of: -Narrator-!npc@[server-name] */
+		std::string narr_source = std::string("-Narrator-!npc@") + ServerInstance->Config->ServerName;
+
+		c->WriteChannelWithServ(narr_source, "PRIVMSG %s :%s%s%s", c->name.c_str(), action ? "\1ACTION " : "", this->text.c_str(), action ? "\1" : "");
+
+		if (localUser)
+		{
+			std::vector<std::string> params;
+			params.push_back("*");
+			params.push_back(this->cmdName);
+			params.push_back(parameters[0]);
+			params.push_back(":" + this->text);
+			ServerInstance->PI->SendEncapsulatedData(params);
+		}
+
+		return CMD_SUCCESS;
+	}
+
+	void SetText(const std::string &newText)
+	{
+		this->text = newText;
+	}
+};
+
+/** Handle /NARRATOR
+ */
+class CommandNarrator : public Command, public Narratorx
+{
+public:
+	CommandNarrator(Module *parent) : Command(parent, "NARRATOR", 2, 2), Narratorx("NARRATOR")
+	{
+		this->syntax = "<channel> <text>";
+	}
+
+	CmdResult Handle(const std::vector<std::string> &parameters, User *user)
+	{
+		return Narratorx::Handle(parameters, user, false);
+	}
+};
+
+/** Handle /NARRATORA
+ */
+class CommandNarratorA : public Command, public Narratorx
+{
+public:
+	CommandNarratorA(Module *parent) : Command(parent, "NARRATORA", 2, 2), Narratorx("NARRATORA")
+	{
+		this->syntax = "<channel> <text>";
+	}
+
+	CmdResult Handle(const std::vector<std::string> &parameters, User *user)
+	{
+		return Narratorx::Handle(parameters, user, true);
+	}
+};
+
+class ModuleRPGCommands : public Module
+{
+	CommandNPC npc;
+	CommandNPCA npca;
+	CommandAmbiance ambiance;
+	CommandNarrator narrator;
+	CommandNarratorA narratora;
+
+public:
+	ModuleRPGCommands() : npc(this), npca(this), ambiance(this), narrator(this), narratora(this)
+	{
+	}
+
+	void init()
+	{
+		ServiceProvider *services[] = { &this->npc, &this->npca, &this->ambiance, &this->narrator, &this->narratora };
+		ServerInstance->Modules->AddServices(services, sizeof(services) / sizeof(services[0]));
+
+		ServerInstance->Modules->Attach(I_OnPreCommand, this);
+	}
+
+	Version GetVersion()
+	{
+		return Version("Provides NPC, NPCA, AMBIANCE, NARRATOR, and NARRATORA commands for use by Game Masters doing pen & paper RPGs via IRC", VF_COMMON);
+	}
+
+	/** The purpose of this is to make it so the command text doesn't require a colon prefixing the text but also to allow a colon anywhere in the line.
+	 */
```

Comment:
As discussed on IRC, the reason for OnPreCommand being used as it is was to prevent a word starting with a colon from having its colon eaten by the command parser. This was mainly because I had people using the commands with smilies (like :D for instance). The issue is only a non-issue on broadcasting because I prefix a colon to the text before sending the ENCAP.